### ISSUE TITLE
is this tfp.bijectors.MatvecLU's weight trainiable?

### ISSUE THREAD

**Issue by MokkeMeguru** [tensorflow/probability#545]
tensorflow==2.0.0rc0
tensorflow-probability==0.8.0rc0

I want to code OpenAI's Glow referr to https://github.com/breadbread1984/glow-flow (but It's Glow is broken. I think ConvolutionInvertible.py is something wrong...).

I tried to use tfp.bijectors.MatvecLU to implement conv1x1 layer. So, I wrote these code. 

```python
# tensorflow==2.0.0rc0
# tensorflow-probability==0.8.0rc0

import tensorflow as tf
import tensorflow_probability as tfp
import tensorflow_probability.python.bijectors as tfb


def trainable_lu_factorization(event_size,
                               trainable,
                               batch_shape=(),
                               seed=None,
                               dtype=tf.float32,
                               name=None):
    with tf.name_scope('trainable_lu_factorization'):
        event_size = tf.convert_to_tensor(event_size,
                                          dtype=tf.int32,
                                          name='event_size')
        batch_shape = tf.convert_to_tensor(batch_shape,
                                           dtype=event_size.dtype,
                                           name='batch_shape')
        random_matrix = tf.Variable(tf.random.uniform(shape=tf.concat(
            [batch_shape, [event_size, event_size]], axis=0),
                                                      dtype=dtype,
                                                      seed=seed),
                                    name='conv1x1_weights')
        random_orthonormal = tf.linalg.qr(random_matrix)[0]
        lower_upper, permutation = tf.linalg.lu(random_orthonormal)
        lower_upper = tf.Variable(
            initial_value=lower_upper,
            trainable=trainable,
            # use_resource=True,
            name='lower_upper')
        return lower_upper, permutation


def main():
    channels = 3
    traininable = True
    conv1x1 = tfb.MatvecLU(
        *trainable_lu_factorization(channels, traininable),
        validate_args=True)
    x = tf.keras.Input(shape=[2, 28, 28, channels])
    # tf.random.uniform(shape=[2, 28, 28, channels])
    fwd = conv1x1.forward(x)
    rev_fwd = conv1x1.inverse(fwd)
    # ==> x

    example_model = tf.keras.Model(inputs=x, outputs=rev_fwd)
    print(example_model.summary())
    # print(tf.reduce_mean(x - rev_fwd))


if __name__ == '__main__':
    main()

# Model: "model_14"
# _________________________________________________________________
# Layer (type)                 Output Shape              Param #   
# =================================================================
# input_15 (InputLayer)        [(None, 2, 28, 28, 3)]    0         
# =================================================================
# Total params: 0
# Trainable params: 0
# Non-trainable params: 0
# _________________________________________________________________
# None
```

I wonder `random_matrix` is correctly traininable? or did I write the wrong code?

Thank you!

**Comment by brianwa84** [comment]
I think you have a couple options in this case.
First, you should probably be using DistributionLambda to capture relevant variables as a proper keras layer.
e.g. 
`fwd = tfd.DistributionLambda(lambda x: conv1x1(tfd.Deterministic(x)))`
`inv = tfd.DistributionLambda(lambda x: tfb.Invert(conv1x1)(tfd.Deterministic(x)))`

I would guess/hope those will get a variable wired in.

But beyond that, I believe may want to be parameterized in terms of `random_matrix`, assuming the next step is to do gradient descent. If you only do gradient descent to `lower_upper`, *I'm guessing* you will exit the manifold of "lu decompositions of orthonormal matrices" (assuming that's important). 
For example:
```python
d=5
with tf.GradientTape() as tape:
  q, _ = tf.linalg.qr(tf.random.normal([d,d]))
  lu, p = tf.linalg.lu(q)
  tape.watch(lu)
  y = tfp.math.lu_solve(lu, p, tf.random.normal([d,d]))
lug=tape.gradient(y, lu)

qp = tfp.math.lu_reconstruct(lu + lug * .01, p)
print(tf.round(tf.matmul(q, q, transpose_b=True) * 10)/10)
print(tf.round(tf.matmul(qp, qp, transpose_b=True) * 10)/10)
```
shows that qp is no longer orthonormal with a finite-difference step.

To get gradients all the way back to `random_matrix`, you could do:
```python

def trainable_lu_factorization(event_size,
                               trainable,
                               batch_shape=(),
                               seed=None,
                               dtype=tf.float32,
                               name=None):
    with tf.name_scope('trainable_lu_factorization'):
        event_size = tf.convert_to_tensor(event_size,
                                          dtype=tf.int32,
                                          name='event_size')
        batch_shape = tf.convert_to_tensor(batch_shape,
                                           dtype=event_size.dtype,
                                           name='batch_shape')
        random_matrix = tf.Variable(tf.random.uniform(shape=tf.concat(
            [batch_shape, [event_size, event_size]], axis=0),
                                                      dtype=dtype,
                                                      seed=seed),
                                    name='conv1x1_weights')
        def lu_p(random_matrix):
          return tf.linalg.lu(tf.linalg.qr(random_matrix).q)
        return tfp.util.DeferredTensor(lambda m: lu_p(m)[0], random_matrix), tfp.util.DeferredTensor(lambda m: lu_p(m)[1], random_matrix, dtype=tf.int32, shape=random_matrix.shape[:-1])
```

Then, putting it all together:
```python
def main():
    channels = 3
    traininable = True
    conv1x1 = tfb.MatvecLU(
        *trainable_lu_factorization(channels, traininable),
        validate_args=True)
    x = tf.keras.Input(shape=[2, 28, 28, channels])
    # tf.random.uniform(shape=[2, 28, 28, channels])
    fwd = tfp.layers.DistributionLambda(lambda x: conv1x1(tfd.Deterministic(x)))
    # tf.Module tracks attribute assignment and monitors for variable or module references.
    fwd.vars = conv1x1.trainable_variables  
    inv = tfp.layers.DistributionLambda(lambda x: tfb.Invert(conv1x1)(tfd.Deterministic(x)))
    inv.vars = conv1x1.trainable_variables
    rev_fwd = inv(fwd(x))
    # ==> x

    example_model = tf.keras.Model(inputs=x, outputs=rev_fwd)
    print(example_model.summary())

    # print(tf.reduce_mean(x - rev_fwd))

main()
```

**Comment by brianwa84** [comment]
Btw another option for parameterizing orthogonal matrices could be
tf.linalg.expm(x - tf.linalg.matrix_transpose(x)), the exponential map from
skew symmetric matrices to orthogonal matrices. Could use this in lieu of
qr(x).q. not sure which would be better for speed or numerics.

**Comment by brianwa84** [comment]
Ref: https://math.stackexchange.com/q/1935577

**Comment by MokkeMeguru** [comment]
Thanks your many help.
But I couldnot reflect your advice to my code...

Here is the my code `https://gist.github.com/MokkeMeguru/09ad03427aea4e0d01970b9aba282e0d`

I applied "Distribution Lambda" to some parts, but it doesn't seem to be able to calculate the gradient...

**Comment by MokkeMeguru** [comment]
I think...
I found Lu decomposition's gradient is not included in tensorflow2.0.0-rc0. [gist](https://gist.github.com/MokkeMeguru/44d90b761d2a2a6e74c6ce08aae0ad62)

Is this one reason this layer doesn't run correctly?

**Comment by MokkeMeguru** [comment]
Or,
But this is not a problem for tensorflow and tensorflow-probability, when I re-read openai's glow, I found [this lines](https://github.com/openai/glow/blob/master/model.py#L494-L500)

Maybe... should we consider gradients between **decomposed matrix** to some outputs not between **original matrix** to them? 

Is tensorflow looking at the **original matrix** ?

**Comment by brianwa84** [comment]
I spoke w/ someone here more familiar w/ GLOW and it seems like people
generally ignore the LU falling off the orthogonal manifold. This would
support going back to a variable storing the LU, and taking grads w/r/t it.

Apparently I was wrong about keras looking inside a tf.Module, but you can
do fwd.vars = bijector.trainable_variables as here
https://colab.sandbox.google.com/gist/brianwa84/4f7ba8091295b30c2dc2fed260deeff2/matveclu.ipynb


Brian Patton | Software Engineer | bjp@google.com

**Comment by MokkeMeguru** [comment]
Thanks your help!
But according to your book, the gradient cannot be calculated for LU decomposition, it seems.

So my new colab is [here](https://colab.research.google.com/gist/MokkeMeguru/31e7daef57d65503fdfc971a87139e10/matveclu.ipynb)

([Original code]
(https://github.com/openai/glow/blob/master/model.py#L494-L500) seems to ignore original matrix to caluculate it's gradient ...)

I think all the problems are completed. 
Thanks a lot!

**Comment by brianwa84** [comment]
Yes, that looks good. You could probably make the permutation
trainable=False, since it's int32.

Brian Patton | Software Engineer | bjp@google.com
